### PR TITLE
Remove unused resources and fix glsl version 

### PR DESCRIPTION
**Pull request by feliwir**

### CODE REVIEW

File: src/ShaderGen/Metal/MetalBackend.cs [hunk 1]
```diff
@@ -478,5 +459,10 @@ internal override string ParameterDirection(ParameterDirection direction)
                     return string.Empty;
             }
         }
+
+        protected override ShaderMethodVisitor VisitShaderMethod(Compilation compilation, string setName, ShaderFunction func)
```
Comment (mellinoe) [inline src/ShaderGen/Metal/MetalBackend.cs]: This is the same as the base version.

**Comment by mellinoe** [comment]
General style remark: Please put spaces between parameters, and try to match the style in general. Running the VS formatter would be a good enough start.

File: src/ShaderGen/LanguageBackend.cs [hunk 2]
```diff
@@ -359,5 +360,40 @@ internal virtual string CorrectCastExpression(string type, string expression)
         {
             return $"({type}) {expression}";
         }
+
+        protected virtual ShaderMethodVisitor VisitShaderMethod(Compilation compilation,string setName,ShaderFunction func)
+        {
+            return new ShaderMethodVisitor(Compilation, setName,func, this);
```
Comment (mellinoe) [inline src/ShaderGen/LanguageBackend.cs]: This function takes a `compilation` parameter but then uses the `Compilation` property. Given how this function is used and overridden currently, I think we should just remove it. It's not doing anything interesting and is doing more harm than good, IMO.

File: src/ShaderGen/ShaderMethodVisitor.cs [hunk 3]
```diff
@@ -347,6 +348,7 @@ public override string VisitIdentifierName(IdentifierNameSyntax node)
                 ResourceDefinition referencedResource = _backend.GetContext(_setName).Resources.Single(rd => rd.Name == symbolName);
                 _resourcesUsed.Add(referencedResource);
                 _shaderFunction.UsesTexture2DMS |= referencedResource.ValueType.Name == "ShaderGen.Texture2DMSResource";
```
Comment (mellinoe) [inline src/ShaderGen/ShaderMethodVisitor.cs]: We should change this line to do an enum comparison as well.

File: src/ShaderGen/ShaderMethodVisitor.cs [hunk 4]
```diff
@@ -337,6 +337,7 @@ public override string VisitIdentifierName(IdentifierNameSyntax node)
             SymbolInfo symbolInfo = GetModel(node).GetSymbolInfo(node);
             ISymbol symbol = symbolInfo.Symbol;
             string containingTypeName = Utilities.GetFullName(symbolInfo.Symbol.ContainingType);
+
```
Comment (mellinoe) [inline src/ShaderGen/ShaderMethodVisitor.cs]: Unnecessary change

File: src/ShaderGen/Glsl/GlslBackendBase.cs [hunk 5]
```diff
@@ -68,56 +68,62 @@ StructureDefinition[] orderedStructures
                 WriteStructure(sb, sd);
             }
 
+            String funcStr, entryStr;
```
Comment (mellinoe) [inline src/ShaderGen/Glsl/GlslBackendBase.cs]: These can be declared inline

File: src/ShaderGen/Glsl/GlslEs300Backend.cs [hunk 6]
```diff
@@ -19,7 +19,7 @@ protected override string CSharpToShaderTypeCore(string fullType)
 
         protected override void WriteVersionHeader(ShaderFunction function, StringBuilder sb)
         {
-            bool useVersion320 = function.UsesTexture2DMS;
+            bool useVersion320 = function.UsesTexture2DMS || function.UsesStructuredBuffer;
```
Comment (mellinoe) [inline src/ShaderGen/Glsl/GlslEs300Backend.cs]: Structured buffers only require #version 310 es

File: src/ShaderGen/LanguageBackend.cs (hunk 2 above)
Comment (mellinoe) [reply src/ShaderGen/LanguageBackend.cs]: nvm -- as you mentioned, derived classes are using different ShaderMethodVisitor subclasses. Still, this should either use the `compilation` parameter passed in, or all of the overloads should just use the `Compilation` field.

**Comment by feliwir** [comment]
Adressed the issues you've mentioned

**Comment by mellinoe** [comment]
Thanks for the fixes, everything looks good now.

**Comment by tgjones** [comment]
@mellinoe I'm still debugging it to find the root cause, but this appears to have broken resource binding, at least for the HLSL backend. After this change, uniform buffer register indices are assigned separately and sequentially, starting from 0, for vertex and fragment shaders, resulting in overlapping registers, which in turn isn't what Veldrid expects.

File: src/ShaderGen/Hlsl/HlslBackend.cs [hunk 7]
```diff
@@ -196,63 +178,43 @@ StructureDefinition[] orderedStructures
 
                 foreach (ResourceDefinition rd in set)
                 {
+                    if (!resourcesUsed.Contains(rd))
```
Comment (tgjones) [inline src/ShaderGen/Hlsl/HlslBackend.cs]: I think this is the problem - it’s necessary to increment the resource index, even if the resource isn’t used.

Comment (tgjones) [reply src/ShaderGen/Hlsl/HlslBackend.cs]: On a related note, I’ve been thinking it would be nice to test against known-good reference outputs. But I wanted to do some PRs to clean up the output formatting before suggesting that.

Comment (mellinoe) [reply src/ShaderGen/Hlsl/HlslBackend.cs]: > it’s necessary to increment the resource index, even if the resource isn’t used.

You're right, that is broken. @feliwir Do you have any time to submit a fix for that?

Comment (feliwir) [reply src/ShaderGen/Hlsl/HlslBackend.cs]: Will do in 2hrs when home from work
